src/System.Net.Http.WinHttpHandler/src/System/Net/Http/WinHttpCookieHelper.cs

```diff
@@ -0,0 +1,95 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+using SafeWinHttpHandle = Interop.WinHttp.SafeWinHttpHandle;
+
+namespace System.Net.Http
+{
+    internal static class WinHttpCookieHelper
+    {
+        private const string CookieHeaderNameWithColon = HttpKnownHeaderNames.Cookie + ":";
+
+        public static void AddResponseCookiesToContainer(WinHttpRequestState state)
+        {
+            HttpRequestMessage request = state.RequestMessage;
+            SafeWinHttpHandle requestHandle = state.RequestHandle;
+            CookieContainer cookieContainer = state.Handler.CookieContainer;
+            
+            Debug.Assert(state.Handler.CookieUsePolicy == CookieUsePolicy.UseSpecifiedCookieContainer);
+            Debug.Assert(cookieContainer != null);
+
+            // Get 'Set-Cookie' headers from response
+            List<string> cookieHeaders =
+                WinHttpResponseParser.GetResponseHeaders(requestHandle, Interop.WinHttp.WINHTTP_QUERY_SET_COOKIE);
+            WinHttpTraceHelper.Trace("WINHTTP_QUERY_SET_COOKIE");
+            
+            foreach (string cookieHeader in cookieHeaders)
+            {
+                WinHttpTraceHelper.Trace(cookieHeader);
+                try
+                {
+                    cookieContainer.SetCookies(request.RequestUri, cookieHeader);
```

Comment:
Will the request.RequestUri be the correct filter URI if automatic redirection is used?